### PR TITLE
feature: Add 'functions_opening_brace' option 'next_line_if_return_typehint'

### PR DESCRIPTION
**Pull request by pmjones**
The `functions_opening_brace` option `next_line_unless_newline_at_signature_end` fixes muti-line function signatures to look like this:

```php
    public function veryLongFoo(
        VeryLongBar $bar,
        VeryLongBaz $baz
    ): VeryLongFoo {
        // function body
    }
```

This PR adds the option `next_line_if_return_typehint` to fix multi-line function signatures so that the opening brace goes on the next line if a return typehint is present:

```php
    public function veryLongFoo(
        VeryLongBar $bar,
        VeryLongBaz $baz
    ): VeryLongFoo
    {
        // function body
    }
```

This has the benefit of being visually similar to a single-line function signature ...

```php
    public function veryLongFoo(VeryLongBar $bar, VeryLongBaz $baz): VeryLongFoo
    {
        // function body
    }
```

... in that the opening brace remains on the next line instead of "jumping" to the same line, and keeps a (mostly) blank line at the top of the function body.

When combined with the `return_type_declaration` option `'space_before' => 'one'` with 4-space indentation, this has the further benefit of aligning the parameter types with the return type:

```php
    public function veryLongFoo(
        VeryLongBar $bar,
        VeryLongBaz $baz
    ) : VeryLongFoo
    {
        // function body
    }
```

That final combination is what I and several teams I have worked with prefer; this PR allows PHP-CS-Fixer to apply it.

### CODE REVIEW

**Comment by pmjones** [comment]
On consideration, a better option name might be merely `next_line`. Let me know if you (reviewer) have a preference.

**Comment by Wirone** [comment]
Hi @pmjones, sorry for lack of response from the core team 🙏. Can you rebase your branch (and resolve conflicts) so we can see the fresh CI result here?

**Comment by pmjones** [comment]
@Wirone rebased and force-pushed!

**Comment by Wirone** [comment]
@pmjones you need to execute `dev-tools/dox.php` in order to regenerate the docs based on the fixers' definition. Then CI should be green 🙂.

**Comment by Wirone** [comment]
@pmjones I'm AFK now, I can check later this evening.

**Comment by pmjones** [comment]
@Wirone Second sticking point: the FixerDefinition CodeSample does not include a typehint, which means either (1) leaving out documentation for `next_line_unless_newline_at_signature_end` entirely; (2) documenting it without using a typehint, which sounds unhelpful; (3) adding a typehint to the code sample and modifying all of the other examples to include that typehint.

None of those sounds great; do you have a preference as to which is the least-awful?

Thanks for your patience!

**Comment by Wirone** [comment]
@pmjones I can't reproduce the problem - `dev-tools/doc.php` works for me on PHP 8.0, 8.1 and 8.2:

image

Did you try `composer update`? 🤷‍♂️ 

Regarding the [second question](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6669#issuecomment-1574064392): I think you can just add new code sample, focused on this change, without modifying existing samples.

Please

**Comment by pmjones** [comment]
Will do, and thanks for your patient assistance!

**Comment by pmjones** [comment]
@Wirone I think I have it -- review again as convenient!

File: tests/Fixer/Basic/CurlyBracesPositionFixerTest.php [hunk 1]
```diff
@@ -379,6 +379,62 @@ function foo()
...
+                }',
+            '<?php
+                function foo(
+                    $bar,
+                    $baz
+                ) : void {
+                }',
+            ['functions_opening_brace' => 'next_line_if_return_typehint'],
+        ];
+
```
Comment (Wirone) [inline tests/Fixer/Basic/CurlyBracesPositionFixerTest.php:437]: This is exactly the same case as `function (next line if return typehint, multiline)`.

File: src/Fixer/Basic/CurlyBracesPositionFixer.php [hunk 2]
```diff
@@ -241,6 +256,27 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
             }
 
             $whitespace = ' ';
+
+            if (self::NEXT_LINE_IF_RETURN_TYPEHINT === $this->configuration[$positionOption]) {
+                $whitespace = $this->whitespacesConfig->getLineEnding().$this->getLineIndentation($tokens, $index);
+
+                $previousTokenIndex = $tokens->getPrevMeaningfulToken($openBraceIndex);
+                $previousToken = $tokens[$previousTokenIndex];
```
Comment (Wirone) [inline src/Fixer/Basic/CurlyBracesPositionFixer.php:264]: `$previousToken` is unused.

File: src/Fixer/Basic/CurlyBracesPositionFixer.php [hunk 3]
```diff
@@ -241,6 +256,27 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
...
+
+                if ($tokens[$previousTokenIndex]->equals(')')) {
+                    if ($tokens[--$previousTokenIndex]->isComment()) {
+                        --$previousTokenIndex;
+                    }
```
Comment (Wirone) [inline src/Fixer/Basic/CurlyBracesPositionFixer.php:269]: Why is this here? Shouldn't we just check if there's `:` + type between `{` (`$openBraceIndex`) and `)`. Any non-meaningful tokens should be skipped. Such decrement looks like optimistic scenario, what if there are 2 comments 😉?

**Review by Wirone — CHANGES_REQUESTED** [review]

File: src/Fixer/Basic/CurlyBracesPositionFixer.php (hunk 3 above)
Comment (pmjones) [reply src/Fixer/Basic/CurlyBracesPositionFixer.php:269]: /me nods along

I confess, I copied-and-pasted the code for `if (self::NEXT_LINE_UNLESS_NEWLINE_AT_SIGNATURE_END)` (starting on line 280) as the basis for this feature. That original code uses this exact same bit of logic around line 288.

My intuition is that if the logic is good in the original location, it's good probably good here -- unless there is some contextual difference that I'm not seeing. Alternatively, if it's *not* good in the original location, perhaps we have identified a bug.

File: src/Fixer/Basic/CurlyBracesPositionFixer.php (hunk 2 above)
Comment (pmjones) [reply src/Fixer/Basic/CurlyBracesPositionFixer.php:264]: You are correct -- will remove!

File: tests/Fixer/Basic/CurlyBracesPositionFixerTest.php (hunk 1 above)
Comment (pmjones) [reply tests/Fixer/Basic/CurlyBracesPositionFixerTest.php:437]: It does appear unnecessary -- will remove.

File: src/Fixer/Basic/CurlyBracesPositionFixer.php [hunk 4]
```diff
@@ -36,6 +36,11 @@ final class CurlyBracesPositionFixer extends AbstractFixer implements Configurab
 {
     use Indentation;
 
+    /**
+     * @internal
+     */
+    public const NEXT_LINE_IF_RETURN_TYPEHINT = 'next_line_if_return_typehint';
```
Comment (Wirone) [inline src/Fixer/Basic/CurlyBracesPositionFixer.php:42]:
```suggestion
    public const NEXT_LINE_IF_RETURN_TYPE_DECLARATION = 'next_line_if_return_type_declaration';
```
Some time ago we decided to get rid of "typehint" convention (see #7338 for example, but there are also deprecated fixers like `function_typehint_space` and `compact_nullable_typehint`), so I believe we should change the name of this option accordingly.

Sorry for long delay, I had tough time recently but slowly coming back to Fixer.

File: src/Fixer/Basic/CurlyBracesPositionFixer.php (hunk 3 above)
Comment (Wirone) [reply src/Fixer/Basic/CurlyBracesPositionFixer.php:268]: Then we should add test case(s) with comments, like `with comment after opening brace (default)` and `with comment after opening brace (same line)`, to ensure it works correctly.

**Review by Wirone — CHANGES_REQUESTED** [review]

File: src/Fixer/Basic/CurlyBracesPositionFixer.php [hunk 5]
```diff
@@ -354,7 +389,7 @@ protected function createConfigurationDefinition(): FixerConfigurationResolverIn
                 ->setDefault(self::SAME_LINE)
                 ->getOption(),
             (new FixerOptionBuilder('functions_opening_brace', 'The position of the opening brace of functions‘ body.'))
-                ->setAllowedValues([self::NEXT_LINE_UNLESS_NEWLINE_AT_SIGNATURE_END, self::SAME_LINE])
+                ->setAllowedValues([self::NEXT_LINE_IF_RETURN_TYPEHINT, self::NEXT_LINE_UNLESS_NEWLINE_AT_SIGNATURE_END, self::SAME_LINE])
```
Comment (keradus) [inline src/Fixer/Basic/CurlyBracesPositionFixer.php:392]: Is the style this proposal brings a part of a standard or used by a bigger (F)OSS community, eg PSR/PER-CS, Symfony, ...?

For us that is a condition to add a rule (or option) to the project.

File: src/Fixer/Basic/CurlyBracesPositionFixer.php [hunk 6]
```diff
@@ -36,6 +36,11 @@ final class CurlyBracesPositionFixer extends AbstractFixer implements Configurab
 {
```
Comment (keradus) [inline src/Fixer/Basic/CurlyBracesPositionFixer.php:36]: btw, rule was renamed, if anything, shall go to Fixer under new name, ref https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7334

**Comment by I70l0teN4ik** [comment]
Would this be added at some point or it's rejected?
Generally I'd also appreciate to have that rule in place. But would expect that one to be simply named: `next_line`.

**Comment by Wirone** [comment]
If it was rejected, it would be closed. It just requires action from the author. Personally I did not have time and motivation to handle this myself, as I had my own PRs and many others from contributors.
